src/Nancy.ViewEngines.Razor/CSharp/CSharpClrTypeResolver.cs

```diff
@@ -0,0 +1,78 @@
+namespace Nancy.ViewEngines.Razor.CSharp
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Web.Razor.Tokenizer.Symbols;
+
+    internal class CSharpClrTypeResolver : ClrTypeResolver<CSharpSymbolType, CSharpSymbol>
+    {
+        public CSharpClrTypeResolver()
+            : base(CSharpSymbolType.Identifier, CSharpSymbolType.Keyword, CSharpSymbolType.Dot, CSharpSymbolType.WhiteSpace)
+        {
+        }
+
+        protected override TypeNameParserStep ResolveType()
+        {
+            var identifier = this.PopFullIdentifier();
+
+            var step = new TypeNameParserStep(identifier);
+
+            if (!this.symbols.Any()) return step;
+
+            if (this.symbols.Peek().Type == CSharpSymbolType.LessThan)

```

Comment:
From here on it needs to be split into well named helper methods so that it is easier to understand what steps it is performing